lib/rust/parser/src/source/code.rs

```diff
@@ -13,98 +13,133 @@ use crate::prelude::*;
 #[derive(Debug, Clone, Default, Eq, PartialEq, Deref)]
 pub struct StrRef<'s>(pub &'s str);
 
-/// A code representation. It can either be a borrowed source code or a modified owned one.
+/// Identifies a location in source code.
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Default,
+    Eq,
+    PartialEq,
+    Serialize,
+    Reflect,
+    Deserialize,
+    PartialOrd,
+    Ord
+)]
+pub struct Location {
+    /// Offset from the beginning, in UTF-8 code units (bytes).
+    #[reflect(hide)]
+    pub utf8:  u32,
+    /// Offset from the beginning, in UTF-16 code units (two-byte words).
+    #[reflect(hide)]
+    pub utf16: u32,
+    /// Line number, starting from 0. The recognized line terminators are CR, LF, or CRLF.
+    #[reflect(hide)]
+    pub line:  u32,
+    /// Offset from start of line, in UTF-16 code units.
+    #[reflect(hide)]
+    pub col16: u32,
+}
+
+impl Add<Length> for Location {
+    type Output = Self;
+
+    fn add(self, rhs: Length) -> Self::Output {
+        Self {
+            utf8:  self.utf8 + rhs.utf8,
+            utf16: self.utf16 + rhs.utf16,
+            line:  self.line + rhs.newlines,
+            col16: if rhs.newlines == 0 { self.col16 } else { 0 } + rhs.line_chars16,
+        }
+    }
+}
+
+/// A code representation.
 #[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Reflect, Deserialize, Deref)]
 #[allow(missing_docs)]
 pub struct Code<'s> {
+    /// The borrowed string data.
     #[serde(serialize_with = "crate::serialization::serialize_cow")]
     #[serde(deserialize_with = "crate::serialization::deserialize_cow")]
     #[reflect(as = "crate::serialization::Code", flatten, hide)]
     #[deref]
-    pub repr:     StrRef<'s>,
-    #[reflect(hide)]
-    offset_utf16: u32,
-    #[reflect(hide)]
-    utf16:        u32,
+    pub repr: StrRef<'s>,
+    #[reflect(flatten)]
+    offset:   Location,
```

Comment:
I would rename the field as well, as it's now more than just the offset.

Reply:
We were overusing "offset"; renamed it to "start".